<launch>
  <!-- debug flags 若是则使用gdb运行 -->
  <!-- GDB: launch-prefix="gnome-terminal -x gdb -ex run \-\-args" -->
  <arg name="debug"                            default="false" />
  <arg name="launch_prefix"                    value="xterm -e gdb -ex run \-\-args" if="$(arg debug)"/>
  <arg name="launch_prefix"                    value="" unless="$(arg debug)"/>

  <!-- args that needs to be passed in from higher-level launch files -->
  <arg name="agent_id"                         default="0"/>
  <arg name="num_robots"                       default="1" />
  <arg name="dimension"                        default="3" />
  <arg name="relaxation_rank"                  default="5"/>
  <arg name="asynchronous"                     default="false"/>
  <arg name="asynchronous_rate"                default="10" />
  <arg name="verbose"                          default="false"/>
  <arg name="RGD_stepsize"                     default="1e-3" />
  <arg name="RGD_use_preconditioner"           default="true" />
  <arg name="RTR_iterations"                   default="3" />
  <arg name="RTR_tCG_iterations"               default="50" />
  <arg name="RTR_gradnorm_tol"                 default="1e-2" />
  <arg name="local_initialization_method"      default="Odometry"/>
  <arg name="update_rule"                      default="Uniform" />
  <arg name="multirobot_initialization"        default="true"/>
  <arg name="acceleration"                     default="false"/>
  <arg name="restart_interval"                 default="50" />
  <arg name="robust_cost_type"                 default="L2" />
  <arg name="GNC_use_probability"              default="true" />
  <arg name="GNC_quantile"                     default="0.9" />
  <arg name="GNC_barc"                         default="5.0" />
  <arg name="GNC_mu_step"                      default="2.0" />
  <arg name="GNC_init_mu"                      default="1e-5" />
  <arg name="robust_opt_num_weight_updates"    default="4" />
  <arg name="robust_opt_num_resets"            default="0" />
  <arg name="robust_opt_min_convergence_ratio" default="0" />
  <arg name="robust_opt_inner_iters_per_robot" default="10" />
  <arg name="robust_init_min_inliers"          default="5" />
  <arg name="max_iteration_number"             default="1000" />
  <arg name="relative_change_tolerance"        default="0.1" />
  <arg name="log_directory"                    default="$(find dpgo_ros)/logs/agent$(arg agent_id)/" />
  <arg name="robot_names_file"                 default="$(find dpgo_ros)/params/robot_names.yaml" />
  <arg name="publish_iterate"                  default="false"/>
  <arg name="visualize_loop_closures"          default="false"/>
  <arg name="complete_reset"                   default="false"/>
  <arg name="enable_recovery"                  default="false"/>
  <arg name="synchronize_measurements"         default="true" />
  <arg name="max_distributed_init_steps"       default="30" />
  <arg name="inter_update_sleep_time"          default="0"/>
  <arg name="weight_convergence_threshold"     default="-1"/>
  <arg name="max_delayed_iterations"           default="0" />
  <arg name="timeout_threshold"                default="15" />

  <!-- 启动节点agent -->
  <node launch-prefix="$(arg launch_prefix)" ns="dpgo_ros_node" name="agent" pkg="dpgo_ros" type="dpgo_ros_node" output="screen">
    <!-- 设置机器人ID -->
    <param name="~agent_id"                         type="int"    value="$(arg agent_id)" />
    <!-- 设置机器人个数 -->
    <param name="~num_robots"                       type="int"    value="$(arg num_robots)"/>
    <!-- 设置维度，必须为3 -->
    <param name="~dimension"                        type="int"    value="$(arg dimension)" />
    <!-- 设置松弛秩，必须不小于维度: 5 -->
    <param name="~relaxation_rank"                  type="int"    value="$(arg relaxation_rank)" />
    <!-- 设置是否异步及其速率: false -->
    <param name="~asynchronous"                     type="bool"   value="$(arg asynchronous)" />
    <param name="~asynchronous_rate"                type="double" value="$(arg asynchronous_rate)" />
    <!-- Rule to select the next robot for update: Uniform -->
    <param name="~update_rule"                      type="str"    value="$(arg update_rule)" />
    <!-- 设置局部初始化方法: Odometry -->
    <param name="~local_initialization_method"      type="str"    value="$(arg local_initialization_method)" />
    <!-- 设置是否跨机器人初始化 -->
    <param name="~multirobot_initialization"        type="bool"   value="$(arg multirobot_initialization)" />
    <!-- 设置RGD优化参数 -->
    <param name="~RGD_stepsize"                     type="double" value="$(arg RGD_stepsize)" />
    <param name="~RGD_use_preconditioner"           type="bool"   value="$(arg RGD_use_preconditioner)" />
    <!-- 设置RTR优化参数 -->
    <param name="~RTR_iterations"                   type="int"    value="$(arg RTR_iterations)" />
    <param name="~RTR_tCG_iterations"               type="int"    value="$(arg RTR_tCG_iterations)" />
    <param name="~RTR_gradnorm_tol"                 type="double" value="$(arg RTR_gradnorm_tol)" />
    <!-- Nesterov加速参数 -->
    <param name="~acceleration"                     type="bool"   value="$(arg acceleration)"/>
    <!-- 设置重启间隔时间: 50 -->
    <param name="~restart_interval"                 type="int"    value="$(arg restart_interval)" />
    <!-- 鲁棒核函数：L2 -->
    <param name="~robust_cost_type"                 type="str"    value="$(arg robust_cost_type)" />
    <!-- GNC是否使用概率: true -->
    <param name="~GNC_use_probability"              type="bool"   value="$(arg GNC_use_probability)" />
    <!-- GNC置信度: 0.9 -->
    <param name="~GNC_quantile"                     type="double" value="$(arg GNC_quantile)" />
    <!-- GNC自行设置阈值 -->
    <param name="~GNC_barc"                         type="double" value="$(arg GNC_barc)" />
    <!-- 其他GNC参数 -->
    <param name="~GNC_mu_step"                      type="double" value="$(arg GNC_mu_step)" />
    <param name="~GNC_init_mu"                      type="double" value="$(arg GNC_init_mu)" />
    <!-- Number of weight updates for robust optimization: 4 -->
    <param name="~robust_opt_num_weight_updates"    type="int"    value="$(arg robust_opt_num_weight_updates)" />
    <!-- Warm start iterate during robust optimization: 0 -->
    <param name="~robust_opt_num_resets"            type="int"    value="$(arg robust_opt_num_resets)" />
    <!-- Minimum ratio of converged weights before terminating robust optimization: 0 -->
    <param name="~robust_opt_min_convergence_ratio" type="double" value="$(arg robust_opt_min_convergence_ratio)" />
    <!-- Number of inner iterations to apply before updating measurement weights during robust optimization / number_robot -->
    <param name="~robust_opt_inner_iters_per_robot" type="int"    value="$(arg robust_opt_inner_iters_per_robot)" />
    <!-- Minimum number of inliers for robust distributed initialization: 5 -->
    <param name="~robust_init_min_inliers"          type="int"    value="$(arg robust_init_min_inliers)" />
    <param name="~verbose"                          type="bool"   value="$(arg verbose)" />
    <!-- Maximum number of iterations: 100 -->
    <param name="~max_iteration_number"             type="int"    value="$(arg max_iteration_number)" />
    <!-- Stopping condition in terms of relative change: 0.1-->
    <param name="~relative_change_tolerance"        type="double" value="$(arg relative_change_tolerance)" />
    <!-- Publish iterate during optimization: false -->
    <param name="~publish_iterate"                  type="bool"   value="$(arg publish_iterate)" />
    <!-- Publish loop closures as ROS markers for visualization: false -->
    <param name="~visualize_loop_closures"          type="bool"   value="$(arg visualize_loop_closures)" />
    <!-- Completely reset dpgo after each distributed optimization round: false -->
    <param name="~complete_reset"                   type="bool"   value="$(arg complete_reset)" />
    <!-- Try to recover and resume optimization after disconnection: false -->
    <param name="~enable_recovery"                  type="bool"   value="$(arg enable_recovery)" />
    <!-- Synchronize shared measurements between robots before each optimization round: true -->
    <param name="~synchronize_measurements"         type="bool"   value="$(arg synchronize_measurements)" />
    <!-- Maximum multi-robot initialization attempts: 30 -->
    <param name="~max_distributed_init_steps"       type="int"    value="$(arg max_distributed_init_steps)" />
    <param name="~inter_update_sleep_time"          type="double" value="$(arg inter_update_sleep_time)" />
    <!-- Threshold for determining measurement weight convergence: -1 -->
    <param name="~weight_convergence_threshold"     type="double" value="$(arg weight_convergence_threshold)" />
    <param name="~max_delayed_iterations"           type="int"    value="$(arg max_delayed_iterations)" />
    <!-- Timeout threshold for considering a robot disconnected: 15 -->
    <param name="~timeout_threshold"                type="double" value="$(arg timeout_threshold)" />
    <param name="~log_output_path"                  type="str"    value="$(arg log_directory)" />
    <rosparam file="$(arg robot_names_file)" />
  </node>


</launch>